livecheck: add --extract-plist

razvanazamfirei opened the pull request:
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [X] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [X] Have you successfully run `brew style` with your changes locally?
- [X] Have you successfully run `brew typecheck` with your changes locally?
- [X] Have you successfully run `brew tests` with your changes locally?

-----
Running `brew livecheck --cask --eval-all` downloads all casks using the `ExtractPlist` strategy. This takes up a significant amount (~20GB) of bandwidth and cache space. 

If this flag is added, it will allow maintainers to check taps more quickly and with less resources consumed.

apainintheneck commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -237,6 +238,18 @@ def run_checks(
 
+        if cask && no_extract_plist && formula_or_cask.livecheck.strategy == :extract_plist
+          skip_info = {
+            cask:     cask.token.to_s,
```
```suggestion
            cask:     cask.token,
```

Nit: Not needed

apainintheneck commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -170,12 +170,13 @@ def resolve_livecheck_reference(
     def run_checks(
       formulae_and_casks_to_check,
       full_name: false, handle_name_conflict: false, check_resources: false, json: false, newer_only: false,
-      debug: false, quiet: false, verbose: false
+      debug: false, quiet: false, verbose: false, no_extract_plist: false
```
Nit: debug, quiet and verbose should probably stay at the end

apainintheneck reviewed (approved):
Out of curiosity I looked it up and this would affect 275 core casks. Outside of memory savings is this faster as well?

Either way this seems reasonable to me.

bevanjkay commented:
> Out of curiosity I looked it up and this would affect 275 core casks. Outside of memory savings is this faster as well?

It is significantly faster to run `brew livecheck` if you skip the `extract_plist`.
The strategy downloads the binary and checks the contents for a version number, whereas most of the time we are checking HTML webpage contents, or a curl response header.
As some of the files are quite large the network requests can take some time to complete. Note that we only use this strategy when all other alternatives are not possible.

I have my own script locally that runs livecheck on a tap after filtering out the `extract_plist` casks, which could be replaced with this flag.

MikeMcQuaid reviewed (changes requested):
> Running `brew livecheck --cask --eval-all` downloads all casks using the `ExtractPlist` strategy. This takes up a significant amount (~20GB) of bandwidth and cache space.

I really appreciate the PR (great work @razvanazamfirei!) but: the above described behaviour makes no sense.

I'd argue instead that it should instead be one of:
- `--extract-plist` is added instead so it's opt-in
- `--eval-all` never attempts to extract plists
- something else where the above behaviour doesn't manifest

reitermarkus commented:
Maybe this could work similarly to `audit` with `--only` and `--except` flags, so it can also be used more generically to validate changes e.g. when changing a strategy.

apainintheneck commented:
> * `--extract-plist` is added instead so it's opt-in

I like this idea. Do we have to worry about backwards compatibility with this command though. If so, we'll have to go through the whole deprecation cycle and the way it's currently written would be the fastest way to get access to it. It seems like more of an internal command so we're probably fine though.

MikeMcQuaid commented:
> Do we have to worry about backwards compatibility with this command though.

100% not.

> It seems like more of an internal command so we're probably fine though.

Agreed.

MikeMcQuaid commented:
We could even perhaps have a documented policy that some/all developer commands do not require deprecation periods at all.

razvanazamfirei commented:
Thank you all for the suggestions! I've reworked this.

1. The default behavior has been changed to skip `ExtractPlist` casks.
2. Single cask checks maintain current behavior.
3. Added `--extract-plist` flag to opt-in to the livecheck strategy.

apainintheneck commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -197,12 +198,11 @@ def run_checks(
                                                          .select { |items| items.length > 1 }
                                                          .flatten
       end
+      formulae_and_casks_total = formulae_and_casks_to_check.count
```
Nit: This is only used in one place so it should be defined near where it's used or removed and just called directly in the if condition.

apainintheneck reviewed (commented):
The code looks good to me now. I left a nit but, of course, you can ignore that.

There does seem to be a problem with the generated man page and shell completions. It's deleting info related to a few commands. I think these were commands that @dduugg moved to the new command DSL recently so maybe that's related somehow. I really have no idea.

I ran the following and it changed some settings locally too but it tried to remove information related to a different command: `brew bump-unversioned-casks`

```console
$ brew generate-man-completions
Writing completions to /usr/local/Homebrew/completions
Manpage and completions updated.
$ git diff --stats
 completions/bash/brew      | 24 ------------------------
 completions/fish/brew.fish | 14 --------------
 completions/zsh/_brew      | 19 -------------------
 docs/Manpage.md            | 21 ---------------------
 manpages/brew.1            | 14 --------------
 5 files changed, 92 deletions(-)
$ git log -1
commit 7b2bfee3638cba1571157f07fd447d9949bd0a6a (HEAD -> master, origin/master, origin/HEAD)
Merge: 0e981bd8f4 74aea8e92d
Author: Kevin <apainintheneck@gmail.com>
Date:   Wed Mar 20 20:40:26 2024 -0700

    Merge pull request #16903 from apainintheneck/prevent-unexpected-network-calls-in-tests

    Prevent unexpected network calls in tests
```

Now I'm even more confused.

MikeMcQuaid commented:
Thanks @razvanazamfirei!

dduugg commented:
> There does seem to be a problem with the generated man page and shell completions. It's deleting info related to a few commands. I think these were commands that @dduugg moved to the new command DSL recently so maybe that's related somehow. I really have no idea.

Thanks for flagging, i'll take a look.

dduugg commented:
> > There does seem to be a problem with the generated man page and shell completions. It's deleting info related to a few commands. I think these were commands that @dduugg moved to the new command DSL recently so maybe that's related somehow. I really have no idea.
> 
> Thanks for flagging, i'll take a look.

Looks like I fat-fingered `Homebrew::DevCmd::BumpUnversionedCask`, should be `Homebrew::DevCmd::BumpUnversionedCasks`. I'll fix that and a test to validate classes map to their file/cmd names as expected.
